---
title: "CHX dependent analysis"
author: "Matteo Zanovello"
date: "2022-01-10"
---

```{r setup, include = False}
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
if (!require("pacman")) install.packages("pacman")
if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
pacman::p_load(tidyverse, data.table, ggrepel, magrittr, devtools, readxl, gprofiler2, janitor, ggpubr, 
               DESeq2, edgeR, annotables, PCAtools, clusterProfiler, "org.Hs.eg.db", enrichplot, readxl, eulerr,
               pcaExplorer, topGO, ggtranscript, "TxDb.Hsapiens.UCSC.hg38.knownGene","phastCons100way.UCSC.hg38",
               "BSgenome.Hsapiens.UCSC.hg38",GenomicRanges,tximport,rlang,
               rtracklayer, ensembldb, biomaRt)

source(here::here("/Users/matteozanovello/Documents/GitHub/splicing_chx",'scripts','salmon_deseq2.R'))
source(here::here("/Users/matteozanovello/Documents/GitHub/splicing_chx",'scripts','make_volcano_plot.R'))

source(here::here("/Users/matteozanovello/Documents/GitHub/splicing_chx",'scripts','splicing_volcanoplot.R'))

#source(here::here("/Users/matteozanovello/Documents/GitHub/splicing_chx",'scripts','GO_analysis_CM.R'))

source(here::here("/Users/matteozanovello/Documents/GitHub/splicing_chx",'scripts','trial_nmd_plots.R'))
#source(here::here("/Users/matteozanovello/Documents/GitHub/splicing_chx",'scripts','ggtranscript.R'))
#source(here::here("/Users/matteozanovello/Documents/GitHub/splicing_chx",'scripts','run_de_irfinder.R'))
```

```{r}
### path to metadata (for salmon and pca)
metadata_dir_upf1 = fread("data/metadata_upf1.csv")
salmon_quant_directory_upf1 = "~/Documents/Github/splicing_chx/data/salmon_upf1/"

### run salmon on different conditions - also print volcano plot here
for (i in c("ctrl","UPF1KD")) {
###for debugging  #mutation = 0 
salmon_deseq2(i, salmon_quant_directory_upf1, metadata_dir_upf1)
###debugging   #res = results_table   #log10padj = 5   #log2FoldCut = 2
bbb <- label_significant(results_table, log2FoldCut = 2, log10padj = 5) + annotate("text", x=0.9, y=0.1, label=i)
plot(bbb)
assign(paste0("DE_volcano_",i),bbb)
assign(paste0("normed_counts_UPF1_",i),normed_counts)
assign(paste0("results_table_UPF1_",i),results_table)
}
#bbb <- label_significant(results_table_UPF1_ctrl, log2FoldCut = 2, log10padj = 100)
#ccc <- label_significant(results_table_UPF1_UPF1KD, log2FoldCut = 2, log10padj = 100)
list_int <- c("TARDBP", "STMN2", "UNC13A", "UNC13B", "UPF1",
              #"PFKP", "AGRN", "RAP1GAP", 
              #"ELAVL3","CELF5", "RANBP1", "NUP188", "CAMK2B", "CDK7",  
              "INSR" #, "CYFIP2", "SYNE1" 
              #"NPY", "VIP"
              )
bbb <- add_name_to_plot(results_table_UPF1_ctrl, list_int)
ccc <- add_name_to_plot(results_table_UPF1_UPF1KD, list_int)
plot(bbb)
plot(ccc)
ggsave(filename = "results/upgrade/de_upf1_cont.png", bbb, height = 5)
ggsave(filename = "results/upgrade/de_upf1_kd.png", ccc, height = 5)
#write.csv(results_table_UPF1_ctrl, "results/results_deseq_ctrlctrl_ctrlTDP43.csv", row.names = F)
#write.csv(results_table_UPF1_UPF1KD, "results/results_deseq_UPF1ctrl_UPF1TDP43.csv", row.names = F)

###salmon in all samples
salmon_deseq2(c("ctrl","UPF1KD"), salmon_quant_directory_upf1, metadata_dir_upf1)
assign(paste0("normed_counts_UPF1_all"),normed_counts)
#assign(paste0("results_table_both"),results_table)
normed_counts_UPF1_long <- normed_counts_UPF1_all[,c(2:17,19)] %>% 
  #dplyr::filter(symbol == "TARDBP" | symbol == "UPF1" | symbol == "INSR") %>%
  pivot_longer(c(1:16)) %>%
  mutate(group = gsub("_[^_]+$", "", name))
```

distribution and pca
```{r}
### gene expression - data distribution and pca analysis

#density plot
normed_counts_UPF1_long %>%
  ggplot(aes(x = value, color = group)) +
  geom_histogram(aes(y=..density..), colour="black", fill="white") +
  geom_density() +
  facet_wrap(facets = vars(group)) +
  scale_color_brewer(palette="Set1") +
  scale_x_log10() +
  labs(x = expression(paste("Lo", g[10], " counts")), y = "Density", color = "Group") +
  theme_classic()

mean_counts_UPF1 <- normed_counts_UPF1_long %>%
  group_by(group, symbol) %>%
  mutate(meanz = mean(value)) %>%
  mutate(varz = var(value)) 

mean_counts_UPF1 %>% 
  ggplot(aes(x=meanz, y=varz, color = group)) +
  geom_point() +
  geom_abline(intercept = 0, slope = 1, color="black") +
  facet_wrap(facets = vars(group)) +
  scale_color_brewer(palette="Set1") +
  scale_x_log10() +
  scale_y_log10() +
  theme_classic()

#quantile-quantile plot
normed_counts_UPF1_long %>%
  ggplot(aes(sample = value, color = group)) +
  geom_qq() +
  geom_qq_line() +
  facet_wrap(facets = vars(group)) +
  scale_color_brewer(palette="Set1") +
  #scale_x_log10() +
  theme_classic()

### pca analysis
pca_upf1 <- normed_counts_UPF1_all[,c(1:17)] %>%
  #arrange(CTRL_ctrl_1) %>% ####can be better?
  distinct(Geneid, .keep_all = TRUE) %>%
  #select(c(1:)) %>%
  column_to_rownames('Geneid')
  #column_to_rownames('ensgene') #%>% t()

meta_df_upf1 <- metadata_dir_upf1 %>% 
  #mutate(condition = factor(condition, levels = c(0,1))) %>%
  #mutate(treatment = factor(treatment, levels = c(0,1))) %>%
  #mutate(group = gsub("_[^_]+$", "", sample)) %>%
    column_to_rownames('sample') 

p_upf1 <- PCAtools::pca(pca_upf1, metadata = meta_df_upf1, removeVar = 0.1)
screeplot(p_upf1, axisLabSize = 18, titleLabSize = 22)
PCAtools::eigencorplot(p_upf1,components = getComponents(p_upf1),
                       metavars = c('condition',"treatment"))
#nice pca plot - PC2/PC3?
biplot(p_upf1,
       x = "PC1",
       y = "PC3",
       colby = 'group')
```

splicing
```{r}
###splicing volcano + slope plot
##delta psi tables
input_upf1 <- c("CtrlCtrl-TDP43Ctrl",
           "CtrlCtrl-CtrlUPF1",
           "TDP43Ctrl-TDP43UPF1",
           "CtrlUPF1-TDP43UPF1")
datalista <- list()
list_gene <- c("UNC13A", "STMN2", "AARS1", #"UNC13B", 
               "CELF5", "ELAVL3", "INSR", "ATG4B", #"KCNQ2", #"RANBP1", 
               "NUP188","SYT7", #"WASL",
                #"RAB1GAP", #"AGRN","PFKP", 
                 #"HDGFL2","ARHGAP32", "CEP290", "PRELID3A", "KALRN",
                 #"CAMK2B", "ADCY1", "TRAPPC12", "EPB41L4A", "IQCE", "SEMA4D"
                 "CYFIP2", "SYNE1")


for (i in input_upf1) {
  input_splicing <- fread(paste0("/Users/matteozanovello/Documents/GitHub/splicing_chx/data/majiq_upf1/", i, "_annotated_junctions.csv"))
  names(input_splicing)[12] <- "base_mean_psi"
  names(input_splicing)[13] <- "contrast_mean_psi"
  assign(paste0("DS_",i), input_splicing)
  datalista[[i]] <- input_splicing
  a <- splicing_dots_tables_function(input_splicing) #+ annotate("text", x=0.7, y=0.1, label=i)
  #plot(a)
  assign(paste0("splicing_volcano_", i), a)
}
big_delta_upf1 <- rbindlist(datalista, idcol = TRUE)
big_delta_upf1$.id <- factor(big_delta_upf1$.id, levels = input_upf1)

xxx <- plot(`splicing_volcano_CtrlCtrl-TDP43Ctrl`)
yyy <- plot(`splicing_volcano_CtrlUPF1-TDP43UPF1`)
plot(xxx)
plot(yyy)

#big_delta_upf1_ctrl <- big_delta_upf1 %>%
#  filter(.id == "CtrlCtrl-CtrlUPF1") #%>%
#  ggplot(aes(x = mean_dpsi_per_lsv_junction, y = -log10(1 - probability_changing), color = de_novo_junctions)) +
#  geom_point() +
#  theme_classic()

#ggsave(filename = "results/upgrade/ds_upf1_ctrl.png", xxx, width = 10, height = 6)
#ggsave(filename = "results/upgrade/ds_upf1_kd.png", yyy, width = 10, height = 6)

##single psi tables
input_list <- c("ctrl_ctrl",
                "ctrl_UPF1", 
                "TDP43_ctrl", 
                "TDP43_UPF1")
datalist <- list()
for (i in input_list_upf1) {
  input_single_splicing <- fread(paste0("/Users/matteozanovello/Documents/GitHub/splicing_chx/data/majiq_upf1/", i, "_parsed.csv"))
  datalist[[i]] <- input_single_splicing
}
big_data_upf1 <- rbindlist(datalist, idcol = TRUE)
big_data_upf1$.id <- factor(big_data_upf1$.id, levels = input_list_upf1)

#slope plot
gene_list <- c("STMN2", "UNC13A", "AARS1")
a_upf1 <- slope_plot_nmd(big_data_upf1, gene_list) ###generalise input_list!!
plot(a_upf1)
```

```{r}
#create input for GO - DE (do it also on DS)
carmm <- results_table_UPF1_ctrl
cm_PR_up <- carmm %>% dplyr::filter(log2FoldChange > 0 & padj < 0.05) 
cm_PR_up <- cm_PR_up[order(-cm_PR_up$log2FoldChange),]
cm_PR_down <- carmm %>% dplyr::filter(log2FoldChange < 0 & padj < 0.05)
cm_PR_down <- cm_PR_down[order(cm_PR_down$log2FoldChange),]

#GO analysis
go_cm_PR <- gprofiler2::gost(query = list(Upregulated = cm_PR_up$symbol,
                                          Downregulated = cm_PR_down$symbol), 
                             sources = c("GO:BP", "GO:CC", "GO:MF"), organism = 'hsapiens', 
                             ordered_query = T, evcodes = TRUE)

go_cm_PR$result %>%
  dplyr::filter(-log10(p_value) > 30) %>%
  arrange(source, desc(query), desc(p_value)) %>%
  mutate(term_code = paste0(term_name, " (", term_id, ")")) %>%
  mutate(term_code=factor(term_code, levels=unique(term_code))) %>%
  ggplot(aes(y = term_code, group = query, x = -log10(p_value), fill = query)) +
  geom_col(position = "dodge", width = 0.8) +
  #geom_point(size = 10, show.legend = F) +
  geom_text(aes(label = intersection_size), 
            nudge_x = -1, size = 0.8) +
  facet_grid(facets = vars(source), 
             scales = "free_y", space = "free_y") +
  theme_classic() +
  scale_fill_manual(values = c("#FAD510", "#3B9AB2")) +
  labs(title = "Enriched GO terms for significant genes in DOXvsCTRL", 
       y = "Term name and code", 
       x = expression(paste("-lo", g[10], " (", italic("P"), " value)")), fill = "") + 
  theme(legend.position = "top",
        legend.justification='left',
        legend.key.size = unit(0.5,"line"),
        strip.text = element_text(
          size = 4)) +
  theme(text = element_text(size = 6)) +
  theme(plot.title = element_text(hjust = -0.5, size = 8)) +
  theme(legend.text=element_text(size=4))


###GO on DS
carm_ds <- big_delta_upf1 %>% 
  dplyr::filter(.id == "CtrlCtrl-TDP43Ctrl" & base_mean_psi < 0.05 & mean_dpsi_per_lsv_junction > 0.1) %>%
  arrange(-mean_dpsi_per_lsv_junction) %>% distinct(gene_name)
go_cm_DS <- gprofiler2::gost(query = list(Upregulated = carm_ds$gene_name),
                                          #Downregulated = cm_DS_down$gene_name), 
                             sources = c("GO:BP", "GO:CC", "GO:MF"), organism = 'hsapiens', 
                             ordered_query = T, evcodes = TRUE)
go_cm_DS$result %>%
  #dplyr::filter(-log10(p_value) > 2) %>%
  arrange(source, desc(query), desc(p_value)) %>%
  mutate(term_code = paste0(term_name, " (", term_id, ")")) %>%
  mutate(term_code=factor(term_code, levels=unique(term_code))) %>%
  ggplot(aes(y = term_code, group = query, x = -log10(p_value), fill = query)) +
  geom_col(position = "dodge", width = 0.8) +
  #geom_point(size = 10, show.legend = F) +
  geom_text(aes(label = intersection_size), 
            nudge_x = -1, size = 0.8) +
  facet_grid(facets = vars(source), 
             scales = "free_y", space = "free_y") +
  theme_classic() +
  scale_fill_manual(values = c("#FAD510", "#3B9AB2")) +
  labs(title = "Enriched GO terms for significant genes in DOXvsCTRL", 
       y = "Term name and code", 
       x = expression(paste("-lo", g[10], " (", italic("P"), " value)")), fill = "") + 
  theme(legend.position = "top",
        legend.justification='left',
        legend.key.size = unit(0.5,"line"),
        strip.text = element_text(
          size = 4)) +
  theme(text = element_text(size = 6)) +
  theme(plot.title = element_text(hjust = 0, size = 8)) +
  theme(legend.text=element_text(size=4))

###term overlap splicing and expression


###venn diagram - compare TDP/CTRL in ctrl vs chx - also compare DE with DS
up_PR_only <- carmm %>% dplyr::filter(FC_PR_WT > 1 & p_value_PR_WT < 0.05 &
                                        (FC_GR_WT < 1 | p_value_GR_WT > 0.05)) 
up_GR_only <- carmm %>% dplyr::filter(FC_GR_WT > 1 & p_value_GR_WT < 0.05 &
                                        (FC_PR_WT < 1 | p_value_PR_WT > 0.05)) 
down_PR_only <- carmm %>% dplyr::filter(FC_PR_WT < 1 & p_value_PR_WT < 0.05 &
                                          (FC_GR_WT > 1 | p_value_GR_WT > 0.05))
down_GR_only <- carmm %>% dplyr::filter(FC_GR_WT < 1 & p_value_GR_WT < 0.05 &
                                          (FC_PR_WT > 1 | p_value_PR_WT > 0.05))
#PR_down 509 - all_down 250 - PR_only 258
#GR_down 474 - all_down 250 - GR_only 223
#PR_up 329 - all_up 214 - PR_only 115
#GR_up 309 - all_up 214 - GR_only 95

mats_down <- c("(GR)400" = 223, "(PR)400" = 258, "(PR)400&(GR)400" = 250)

fit_down <- euler(mats_down)
#pdf("Venn_down.pdf", width = 3.5, height = 2, pointsize = 12)
plot(fit_down, main = "Venn diagram showing downregulated genes in the two models",
     quantities = list(type = c("counts")), labels = FALSE,
     fills = c("#ffa040", "#55a0fb"))
#dev.off()

mats_up <- c("(GR)400" = 95, "(PR)400" = 115, "(PR)400&(GR)400" = 214)
fit_up <- euler(mats_up)
#pdf("Venn_up.pdf", width = 3.5, height = 2, pointsize = 12)
plot(fit_up, main = "Venn diagram showing upregulated genes in the two models",
     quantities = list(type = c("counts")), labels = FALSE,
     fills = c("#ffa040", "#55a0fb"))
#dev.off()

```

#ggtranscript visualization
```{r}
postar_bedd <- fread("~/Desktop/rbp_bed/human_RBP_binding_sites_sorted.bed", 
                    col.names = c("seqnames", "start", "end", "dataset", "score", "strand", "QC"))

#sy5y_tdp_bed_1 <- fread("~/Desktop/rbp_bed/tardbp-shsy5y-1-20210701-mh_mapped_to_genome_single_peaks.bed", 
#                      col.names = c("seqnames", "start", "end", "dataset", "score", "strand"))
#sy5y_tdp_bed_1 <- sy5y_tdp_bed_1 %>%
#  mutate(QC = "-") %>%
#  mutate(dataset = "zSy5yTDP")

#sy5y_tdp_bed_2 <- fread("~/Desktop/rbp_bed/tardbp-shsy5y-2-20210701-mh_mapped_to_genome_single_peaks.bed", 
#                        col.names = c("seqnames", "start", "end", "dataset", "score", "strand"))
#sy5y_tdp_bed_2 <- sy5y_tdp_bed_2 %>%
#  mutate(QC = "-") %>%
#  mutate(dataset = "zSy5yTDP")

#postar_bed <- rbind(postar_bedd, sy5y_tdp_bed_1, sy5y_tdp_bed_2)
postar_bed <- postar_bedd

ensembl <- biomaRt::useEnsembl(biomart = "genes", dataset = "hsapiens_gene_ensembl")
humandb <- biomaRt::getBM(attributes = c("external_gene_name", 
                                         "ensembl_gene_id", "ensembl_transcript_id", "transcript_appris", 
                                         "uniprotswissprot", "uniprotsptrembl", "uniprot_isoform"), 
                          mart = ensembl)
princ <- humandb[which(humandb$transcript_appris == "principal1" |
                       humandb$transcript_appris == "principal2" | 
                       humandb$transcript_appris == "principal3" | 
                       humandb$transcript_appris == "principal4" |
                       humandb$transcript_appris == "principal5"), ]

cds_regions = cdsBy(TxDb.Hsapiens.UCSC.hg38.knownGene, "tx",use.names = TRUE)
cds_regions = unlist(cds_regions)
cds_regions$transcript_id = gsub("\\..*", "", names(cds_regions))

exons_regions = exonsBy(TxDb.Hsapiens.UCSC.hg38.knownGene, "tx",use.names = TRUE)
exons_regions = unlist(exons_regions)
exons_regions$transcript_id = gsub("\\..*", "", names(exons_regions))

#file_path = "data/sj_tabs/normalized_annotated/"
#suffix = "_normalized_annotated.csv"
#psi_files = list.files(file_path,
#                       pattern = suffix,
#                       full.names = TRUE)
#psi_list_full = purrr::map(psi_files,my_clean_reader)
#samp_ids = base::tolower(purrr::simplify(purrr::map(psi_files, basename)))
#samp_ids = gsub(suffix,"",samp_ids)
##add on the name as an additional column
#psi_list_full = purrr::map2(psi_list_full, samp_ids, ~cbind(.x, SampleID = .y))
#psi_list_full = purrr::map(psi_list_full,my_name_fixer)
#psi_list_full = data.table::rbindlist(psi_list_full)

gene_list <- c("NTRK2", "UNC13A", "STMN2")
source(here::here('scripts','ggtranscript.R'))
for (i in gene_list) {
 transcript_bind_plot(i)
  }
```

IRFinder
```{r}
ir_top_level = "data/irfinder/"
experiment = make_deframe_irfinder(ir_top_level,base_grep = "CTRL",
                                   contrast_grep = "DOX")$metadata
paths = make_deframe_irfinder(ir_top_level,base_grep = "CTRL",
                                   contrast_grep = "DOX")$path

#for (j in c("0.0125", "0.0187", "0.021", "0.025", "0.075")) {
    experiment <- dplyr::filter(experiment, grepl("ctrl", SampleNames))
    paths <- paths[grepl("ctrl", paths)]
#}

dds = run_deseq_ir(paths,experiment = experiment)
result = return_formated_results(dds)
resultCTRL <- result %>% 
    mutate(ir_change = TDPKD -  control)
```
